### DISCUSSION TITLE
Add ability to use Copy with file-like objects

### DISCUSSION THREAD

**Discussion by sigmavirus24** [psycopg/psycopg#111] [Ideas]
I understand and appreciate that the new Copy interface is far better for folks who want to do things in memory, but it's not clear that it's as efficient as generating a file and the old interface. Some folks have a fairly good set of performance characteristics with that in psycopg2 and I'm wondering if there's any willingness to support files/file-like objects as well to make the upgrade experience easier and nicer

**Comment by sigmavirus24** [comment]
Also, I'd be happy to contribute this if there's a willingness to be open to adding this

**Comment by dvarrazzo** [comment]
> I understand and appreciate that the new Copy interface is far better for folks who want to do things in memory

why in memory? Folks using it keep a single block in memory. As psycopg2 does, internally...

> but it's not clear that it's as efficient as generating a file and the old interface

Can you show numbers proving that it is less efficient?

The psycopg2 implementation is no more than a loop which either passes the block of data returned by COPY TO  to `file.write()` or reads a block from `file.read()` and passes it to a COPY FROM. There is no specialisation in case the file is a real file and not a file-like object. So, if even there was a difference in performance, I doubt it would be solved by introducing a function that takes a file as argument. If you are seeing a difference in performance I'm happy to look into that.

As to provide a wrapper function emulating `copy_expert()`: it's easy stuff but I'm not sure I see the appeal, and I talk as someone who has used psycopg2 for 15 years. `copy_expert()` is quite ill defined: you don't know if it does input or output, it depends on the query, and `copy_from/to()` take a set of parameters [designed on PostgreSQL 7.3](https://www.postgresql.org/docs/7.3/sql-copy.html). I am quite pleased to get rid of this baggage, and Psycopg 3 is a great chance to do so.

> Also, I'd be happy to contribute this if there's a willingness to be open to adding this

What, specifically, you would like to see implemented or contribute to? Are we talking of difference in performance or to introduce the old interface? Which function/s?

**Reply by sigmavirus24** [reply]
>  If you are seeing a difference in performance I'm happy to look into that.

We saw the announcement and started perusing the code. We haven't yet tested out 3.0. Given the size of the files we generate we're concerned about the throughput of a `queue.Queue` as is used in https://github.com/psycopg/psycopg/blob/2a583c9c4432893e5ebfe9567a0928665a619fde/psycopg/psycopg/copy.py#L274-L284 we have significant experience around these mechanisms and know that with enough throughput they will start to slow down our application's performance.

> So, if even there was a difference in performance, I doubt it would be solved by introducing a function that takes a file as argument.

So the old method (`copy_expert`) was in C and eventually lands in https://github.com/psycopg/psycopg2/blob/1d3a89a0bba621dc1cc9b32db6d241bd2da85ad1/psycopg/pqpath.c#L1204 which doesn't have any queue buffering whatsoever. I fully understand that https://github.com/psycopg/psycopg2/blob/1d3a89a0bba621dc1cc9b32db6d241bd2da85ad1/psycopg/pqpath.c#L1232 is just reading the file in chunks but based on my own copious experience around Python performance, a C for-loop (even one written with `Python.h`) will far surpass the pure python, especially pure-python involving threads and queues.

I can work on benchmarks, but I'm surprised that this is even a debate.

> I am quite pleased to get rid of this baggage

I totally get wanting to get rid of maintenance burden. I want to find a way to keep the performance characteristics and allow us to side-step the overhead of threads, queues, and in-memory buffers so we can just pass in the prepared file objects for this.

> What, specifically, you would like to see implemented or contribute to?

Effectively the essence of what we want is high-performance "here's a file, tell Postgres to copy it" interface. It doesn't have to exactly be `copy_expert`. I'd settle for something in Cython that can match the old performance or something near it though

**Reply by dvarrazzo** [reply]
>  we have significant experience around these mechanisms and know that with enough throughput they will start to slow down our application's performance.

If you could test and let us know about regression performances, or even better to fix them, that would be great.

> Given the size of the files we generate we're concerned about the throughput of a `queue.Queue`...

That queue size is limited to 1024 items, which will be 32K bytes strings, so as big as your file will be it will use no more than 32Mb and a bit of fat. Using that query and a thread decouples the I/O-bound part of copying (sending data down the socket) from the CPU-bound part (converting Python objects to bytes). In my test it gave some good performance improvement than using a single thread doing alternatively CPU work and then blocking I/O work. 

If you copy from a file rather than from Python objects there will be less CPU involved but it's still two data sources being read and written in parallel in two different threads, at least until the queue is maxed out.

Happy if you take a critical look at the code though and work out if there are better alternative. Or if there are stupid errors, such as forgetting to release the GIL around some I/O operation.

> based on my own copious experience around Python performance, a C for-loop (even one written with Python.h) will far surpass the pure python, especially pure-python involving threads and queues.

> I can work on benchmarks, but I'm surprised that this is even a debate.

It is for me, and I'd rather see numbers. If we were multiplying matrices I would understand, but here we are in the middle of a pretty I/O intense operation, and we are talking to add an the amount of opcodes that:
```
        for block in copy:
            write(block)
```
generates for every 8k of data shipped or received on the network. I still think that this will be pretty I/O bound...

> Effectively the essence of what we want is high-performance "here's a file, tell Postgres to copy it" interface. It doesn't have to exactly be copy_expert. I'd settle for something in Cython that can match the old performance or something near it though

If there is a consistent performance improvement to gain from a C implementation I would be happy to introduce it.

**Reply by sigmavirus24** [reply]
> It is for me, and I'd rather see numbers.

I'll happily work on those. Any constraints you think are important to keep in mind? (Don't want to do a bunch of work and have it thrown out because I missed something that might be obvious to you)

**Reply by dvarrazzo** [reply]
So, with copy, of course, performance matter, and I am interested in add features as long as they are actually beneficial and don't become a maintenance burden for no upside. All in all, having a file based copy doesn't seem extremely difficult to introduce. I would say:

- Can you provide a benchmark of how faster would file-based copy would be? 
- what improvement threshold, in your opinion, would be worth to introduce a new interface? 20%? (yeah sure!) 1%? (I don't know...)
- what interface to give it? I'd say a pair of method `read_file()` to consume a file for COPY... FROM STDIN and a `write_file()` to COPY... TO STDOUT

If you could quantify how much a file based interface would be better than a trivial loop in Python, and it turns out to be yummy, happy to work on it, or to receive your contribution :)

Thank you very much!
